### PR TITLE
Config dev

### PR DESCRIPTION
**Pull request by Jake-Janssen**

### CODE REVIEW

File: pcs_detection/scripts/data/aluminum_3/training_labels.xml [hunk 1]
```diff
@@ -0,0 +1,181 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (mpowelson) [inline pcs_detection/scripts/data/aluminum_3/training_labels.xml]: Can we rename this dataset to example_dataset_1 or something similar. That's why I had those dataset directories. Also, could we split them into 2 separate datasets to show how to combine multiple?

Comment (mpowelson) [reply pcs_detection/scripts/data/aluminum_3/training_labels.xml]: It doesn't have to be anything special. You could just copy the second half into the other directory

File: pcs_detection/scripts/data/fcn8_train.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ '''
+
+import json
+import os        
+from src_python.pcs_detection.process import test_dataloader, train
```
Comment (mpowelson) [inline pcs_detection/scripts/data/fcn8_train.py]: I realize you are not installing this with cmake, but when you do it will install the pcs_detection module without you having to do this. In fact, I'm a bit surprised this works unless you are doing something tricky with your Python path.

Long story, short please change `src_python.pcs_detection.process` to `pcs_detection.process`

File: pcs_detection/scripts/data/fcn8_validate.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class Config:
+    def __init__(self, **entries):
+        self.__dict__.update(entries)
+
+if __name__ == '__main__':
+    # Import Config json file and convert into format we need
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    with open(dir_path + '/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json') as json_data_file:
```
Comment (mpowelson) [inline pcs_detection/scripts/data/fcn8_validate.py]: Can we rename this to something more reasonable? Like `example_weights` or `example_weld_detector_weights`?

File: pcs_detection/scripts/fcn8_inference.py [hunk 4]
```diff
@@ -34,7 +34,7 @@
 import os
 
 # Inference module
-from pcs_detection.inference import Inference
+from src_python.pcs_detection.inference import Inference
```
Comment (mpowelson) [inline pcs_detection/scripts/fcn8_inference.py]: See above

File: pcs_detection/src_python/pcs_detection/utils.py [hunk 5]
```diff
@@ -42,7 +42,7 @@ def get_labels_from_xml(label_path):
     
     # name of the outer directory this label corresponds to
     # allows for images in different directories to have the same filename 
-    dataset_name = label_path.split('/')[4]
+    dataset_name = label_path.split('/')[-2]
```
Comment (mpowelson) [inline pcs_detection/src_python/pcs_detection/utils.py:45]: Is -2 always correct?

File: pcs_detection/scripts/data/weights/.gitignore [hunk 6]
```diff
@@ -1,4 +1,2 @@
-*
```
Comment (mpowelson) [inline pcs_detection/scripts/data/weights/.gitignore:1]: I believe this is needed to keep it from saving the weight file is it not?

File: pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        "reduce_factor": 0.6
+    },
+    "N_EPOCHS": 100,
+    "IMG_DIMS": [
+        256,
+        256
+    ],
+    "USE_FULL_IMAGE": false,
+    "VAL_WEIGHT_PATH": "38.h5",
+    "TRAINING_DIRS": [
```
Comment (mpowelson) [inline pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json]: Config is for infererence isn't it? Then it shouldn't need the training data, validation_data, learning rate, or background_reduction.

File: pcs_detection/scripts/data/training_config.json [hunk 8]
```diff
@@ -0,0 +1,69 @@
+{
+    "MODE": "TRAIN",
+    "WEIGHT_ID": "new_repo_test",
+    "WEIGHT_DIR": "data/weights",
```
Comment (mpowelson) [inline pcs_detection/scripts/data/training_config.json]: I would think this directory is relative to the location of the config? That doesn't look correct in this case. /scripts/data/data/weights isn't correct.

File: pcs_detection/scripts/data/fcn8_validate.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env python3
+'''
+ * @file fcn8_validate.py
```
Comment (mpowelson) [inline pcs_detection/scripts/data/fcn8_validate.py]: These should be in scripts not data.

File: pcs_detection/scripts/data/training_config.json [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
+    "MODEL": "fcn8",
+
+    "CHANNEL": "GREY",
```
Comment (mpowelson) [inline pcs_detection/scripts/data/training_config.json]: Could you create a readme in pcs_detection that describes these configs?

File: pcs_detection/scripts/data/training_config.json [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+            117
+        ]
+    },
+    
+    "USE_FULL_IMAGE": false,
+    "DISPLAY_SCALE_FACTOR": 1.25,
+
+    "TRAINING_DIRS": [
+        {
+            "dir_path": "data/aluminum_3/training",
```
Comment (mpowelson) [inline pcs_detection/scripts/data/training_config.json]: These paths are relative to what?

**Review by mpowelson — CHANGES_REQUESTED** [review]
The main thing that needs to be fixes is the imports. You may need to add pcs_detection to your path manually to get it to work, but a a person building this in cmake would have it installed at compile time.

Also, the configs don't look that cleaned up. I'd like to see the inference config be pretty simple. It shouldn't need paths to datasets or anything like that. All I'd imagine it needing is a path to the weights and info on what preprocessing to do.

File: pcs_detection/scripts/data/fcn8_train.py (hunk 2 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/fcn8_train.py]: Ah that makes sense, I could not get the import from pcs_detection.process to work so I ran it from the outer pcs_detection directory

File: pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json (hunk 7 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json]: That's true, none of those parameters are needed to run. I typically save off everything the weights were trained on as a reference for future training.

Comment (mpowelson) [reply pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json]: Yeah. Maybe you could dump that into a log file or something. Or even store it in the h5 file. It just seems a bit confusing what are actual inputs.

File: pcs_detection/src_python/pcs_detection/utils.py (hunk 5 above)
Comment (Jake-Janssen) [reply pcs_detection/src_python/pcs_detection/utils.py:45]: Should be with the current convention. I typically like to leave labels.xml in a directory above the actual images

Comment (mpowelson) [reply pcs_detection/src_python/pcs_detection/utils.py:45]: If that is a requirement we should put that in the readme.

File: pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json (hunk 7 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json]: That makes sense, I'll put it in a log file

File: pcs_detection/scripts/data/training_config.json (hunk 11 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/training_config.json]: This is relative to being in the outer pcs_detection directory and running python3 scripts/data/fcn8_train.py
When using cmake are you able to run it from the data directory?

Comment (mpowelson) [reply pcs_detection/scripts/data/training_config.json]: I would say these paths need to be relative to the config file path. I have not tried to run it, but I would expect the scripts to go in /scripts not /scripts/data. Then the code should run regardless of where you are running it from.

File: pcs_detection/scripts/data/fcn8_validate.py (hunk 3 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/fcn8_validate.py]: renamed to example_weights

File: pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json (hunk 7 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/weights/using_negatives_fcn8_GREY_19_10_10_101037/config.json]: Two files are now saved off after training. One is the full training config which is used for the example_weights.py since data_loader.py requires most parameters for label reconstruction. I also made an inference_config that has only requirements for inference

File: pcs_detection/scripts/data/training_config.json (hunk 8 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/training_config.json]: training config and script moved to scripts directory

File: pcs_detection/scripts/data/training_config.json (hunk 10 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/training_config.json]: Added readme with descriptions of each config and an explanation for all config keys

File: pcs_detection/scripts/data/training_config.json (hunk 11 above)
Comment (Jake-Janssen) [reply pcs_detection/scripts/data/training_config.json]: config moved up a directory

File: pcs_detection/src_python/pcs_detection/utils.py (hunk 5 above)
Comment (Jake-Janssen) [reply pcs_detection/src_python/pcs_detection/utils.py:45]: This will be a requirement, which readme should it go in? The general one or the one specific to pcs_detection

Comment (mpowelson) [reply pcs_detection/src_python/pcs_detection/utils.py:45]: pcs_detection
